### PR TITLE
Reject calls to AddPendingEntry() if etcd has too many entries.

### PR DESCRIPTION
**Pull request by AlCutter**

### CODE REVIEW

File: cpp/log/etcd_consistent_store-inl.h [hunk 1]
```diff
@@ -789,6 +879,79 @@ util::StatusOr<int64_t> EtcdConsistentStore<Logged>::CleanupOldEntries() {
 }
 
 
+template <class Logged>
+void EtcdConsistentStore<Logged>::StartEtcdStatsFetch() {
+  if (etcd_stats_task_.task()->CancelRequested()) {
+    etcd_stats_task_.task()->Return(
+        util::Status(util::error::CANCELLED, "Cancelled."));
```
Comment (pphaneuf) [inline cpp/log/etcd_consistent_store-inl.h]: A nicer message is kind of nice, but I'd stick to just ```util::Status::CANCELLED```, this is a pattern that happens a lot, best to keep it low-profile.

File: cpp/log/etcd_consistent_store-inl.h [hunk 2]
```diff
@@ -789,6 +879,79 @@ util::StatusOr<int64_t> EtcdConsistentStore<Logged>::CleanupOldEntries() {
...
+      bind(&EtcdConsistentStore<Logged>::EtcdStatsFetchDone, this, response,
+           std::placeholders::_1)));
+  client_->GetStoreStats(response, stats_task);
+}
+
+
+template <class Logged>
+void EtcdConsistentStore<Logged>::EtcdStatsFetchDone(
+    EtcdClient::StatsResponse* response, util::Task* task) {
+  CHECK_NOTNULL(response);
```
Comment (pphaneuf) [inline cpp/log/etcd_consistent_store-inl.h]: Could inline that one in the deleter?

File: cpp/log/etcd_consistent_store-inl.h [hunk 3]
```diff
@@ -789,6 +879,79 @@ util::StatusOr<int64_t> EtcdConsistentStore<Logged>::CleanupOldEntries() {
...
+    if (num_entries.ok()) {
+      {
+        std::lock_guard<std::mutex> lock(mutex_);
+        num_etcd_entries_ = num_entries.ValueOrDie();
+      }
+      etcd_total_entries->Set("all", num_etcd_entries_);
+    } else {
+      VLOG(1) << num_entries.status();
+    }
+  }
```
Comment (pphaneuf) [inline cpp/log/etcd_consistent_store-inl.h]: Maybe log something if the task wasn't OK? Shouldn't happen much, right?

File: cpp/log/etcd_consistent_store-inl.h [hunk 4]
```diff
@@ -789,6 +879,79 @@ util::StatusOr<int64_t> EtcdConsistentStore<Logged>::CleanupOldEntries() {
...
+    const util::StatusOr<int64_t> num_entries(
+        CalculateNumEtcdEntries(response->stats));
+    if (num_entries.ok()) {
+      {
+        std::lock_guard<std::mutex> lock(mutex_);
+        num_etcd_entries_ = num_entries.ValueOrDie();
+      }
+      etcd_total_entries->Set("all", num_etcd_entries_);
+    } else {
+      VLOG(1) << num_entries.status();
```
Comment (pphaneuf) [inline cpp/log/etcd_consistent_store-inl.h]: That's going to be a bit bare, if it happens (maybe as little as "UNKNOWN: "!).

File: cpp/log/etcd_consistent_store-inl.h [hunk 5]
```diff
@@ -789,6 +879,79 @@ util::StatusOr<int64_t> EtcdConsistentStore<Logged>::CleanupOldEntries() {
...
+// This method attempts to modulate the incoming traffic in response to the
+// number of entries currently in etcd.
+//
+// Once the number of entries is above reject_threshold, we will start
+// returning a RESOURCE_EXHAUSTED status, which should result in a 503 being
+// sent to the client.
+template <class Logged>
+util::Status EtcdConsistentStore<Logged>::MaybeReject(
+    const std::string& type) const {
+  if (!cluster_config_) {
```
Comment (pphaneuf) [inline cpp/log/etcd_consistent_store-inl.h]: Doesn't that need to be protected by the lock?

File: cpp/log/etcd_consistent_store-inl.h [hunk 6]
```diff
@@ -84,8 +148,13 @@ template <class Logged>
 EtcdConsistentStore<Logged>::~EtcdConsistentStore() {
   VLOG(1) << "Cancelling watch tasks.";
   serving_sth_watch_task_.Cancel();
+  cluster_config_watch_task_.Cancel();
   VLOG(1) << "Waiting for watch tasks to return.";
   serving_sth_watch_task_.Wait();
+  cluster_config_watch_task_.Wait();
+  VLOG(1) << "Cancelling stats task.";
+  etcd_stats_task_.Cancel();
```
Comment (pphaneuf) [inline cpp/log/etcd_consistent_store-inl.h]: Would be slightly better to cancel everything, then wait on everything?

This has a slight smell that maybe we should just have one SyncTask, and use child tasks for all these things, but don't do this in this PR, just putting the idea in your head... :wink:

**Comment by pphaneuf** [comment]
Done for this round.

File: cpp/log/etcd_consistent_store-inl.h (hunk 1 above)
Comment (AlCutter) [inline cpp/log/etcd_consistent_store-inl.h]: I forgot about the static `CANCELLED` instance, thanks.

File: cpp/log/etcd_consistent_store-inl.h (hunk 2 above)
Comment (AlCutter) [inline cpp/log/etcd_consistent_store-inl.h]: It causes the line to wrap so I'm not sure there's much to be gained...

File: cpp/log/etcd_consistent_store-inl.h (hunk 3 above)
Comment (AlCutter) [inline cpp/log/etcd_consistent_store-inl.h]: Done.

File: cpp/log/etcd_consistent_store-inl.h (hunk 4 above)
Comment (AlCutter) [inline cpp/log/etcd_consistent_store-inl.h]: Added verbiage.

File: cpp/log/etcd_consistent_store-inl.h (hunk 5 above)
Comment (AlCutter) [inline cpp/log/etcd_consistent_store-inl.h]: Whoops, yes.

File: cpp/log/etcd_consistent_store-inl.h (hunk 6 above)
Comment (AlCutter) [inline cpp/log/etcd_consistent_store-inl.h]: Yeah, good call.
Another day, another PR.

File: cpp/log/etcd_consistent_store-inl.h (hunk 2 above)
Comment (pphaneuf) [inline cpp/log/etcd_consistent_store-inl.h]: Okay, fine like this, then.

**Comment by pphaneuf** [comment]
LGTM

**Comment by AlCutter** [comment]
Fixes #606

**Pull request merged**